### ISSUE TITLE
Can I disable auto-insertion of space when hitting tab key?

### ISSUE THREAD

**Issue by aiotter** [akinomyoga/ble.sh#226]
When I hit a tab key with only one completion available, ble.sh executes completion immediately, and inserts space after completed word.

```bash
$ ls -a
.       ..      abc.txt

$ cat ab  # <- press tab key here

$ cat abc.txt_  # <- completed with whitespace (shown as _ here)
```

The same things happens when you hit return on completion list.
I don't like this behaviour. Is it possible to disable this?

**Comment by akinomyoga** [comment]
> I don't like this behaviour. Is it possible to disable this?

Thank you. If the candidate is generate by the programmable completion, you may specify a complete option `nospace` to the `complete` setting or run `compopt -o nospace` in the completion function. But this only applies to the completion candidates generated by the programmable completion, and another problem is that we need to specify the completion option `nospace` to each completion setting one by one. Currently, there is no global option for this to completely turn off suffixing the space.

Now, I hesitate to provide the global setting as a user option. As far as I try, not only the built-in completion of `bash` but also `zsh` and `fish` add a space after a unique completion by the default configuration, though I haven't checked if `zsh` and `fish` have even a user option to turn off the space. Would you think the feature to turn it off should be provided as a user option? I currently see not so much benefits in providing customizability for this behavior maybe because I do not see the actual use cases.

I'm not sure if this is what you want, but in theory it is of course possible to change the behavior if you overwrite the related functions of ble.sh. The suffixing is performed by the function [`ble/complete/action/complete.addtail`](https://github.com/akinomyoga/ble.sh/blob/793dfadf3524adbb48145185493be7be73b12d08/lib/core-complete.sh#L1170-L1172). You may overwrite the function in [the `complete_load` hook](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A77-Completion#7-completion). For example,


```bash
# blerc

function my/overwrite-addtail {
  # Overwrite the function in the hook
  function ble/complete/action/complete.addtail {
    # do nothing
    :
  }
}

blehook/eval-after-load complete my/overwrite-addtail
```

Please not that all the suffixing including `/` after the directory name, `=` after the variable name, etc. are performed by this function, so you might want to disable the suffixing conditionally in `ble/complete/action/complete.addtail` depending on what behavior you like for each case.

**Comment by aiotter** [comment]
Thanks for detailed description! I confirmed that zsh has no options as well, and I agree that this settings is not worth implementing as a user option. The reason why I need it is not a practical one, but just a feeling of uncomfortableness of having unnecessary spaces at the end of all entries of command histories.

Following what you suggested, I successfully append the following settings to solve this problem:

```bash
      function my/dont-add-space-on-tail {
        function ble/complete/action/complete.addtail {
          [[ "$1" != " " ]] && suffix=$suffix$1
        }
      }
      blehook/eval-after-load complete my/dont-add-space-on-tail
```

Thank you for taking time!

**Comment by akinomyoga** [comment]
> The reason why I need it is not a practical one, but just a feeling of uncomfortableness of having unnecessary spaces at the end of all entries of command histories.

That made sense to me, so I have experimentally implemented in 021e0330d127c254560976ac208c0b39ecebc2dd an extension `HISTCONTROL=…:strip`, where the leading and trailing spaces of the command line are stripped before the command line is saved in the command history. I'd appreciate it if you could try it and let me know how you feel. If you like it, maybe I also submit [a patch for the native support for `strip`](https://gitlab.com/akinomyoga/bash/-/commit/d43c167e9ec150d1fe4a730475a590d48a4cc9cd) to the upstream Bash.

**Comment by aiotter** [comment]
Brilliant! I tried it and this is awesome. I'm really happy to see the best approach for this.
